### PR TITLE
proc,service: pretty print time.Time variables

### LINKED ISSUE #999: Add special formatting for time.Time
Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.

1. What version of Delve are you using (`dlv version`)?
Delve Debugger
Version: 1.0.0-rc.2
2. What version of Go are you using? (`go version`)?
go version go1.9 windows/amd64
3. What operating system and processor architecture are you using?
Windows10 64bit
4. What did you do?

``` golang
package main

import (
    "fmt"
    "time"
)
func main() {
    t:=time.Now()
    fmt.Println(t)
}
```

and run 
```bash
dlv debug
b main.main
c
...
print t
```

5. What did you expect to see?
I want to get string like below formatted time text
`2017-10-23 11:03:00 ...`
6. What did you see instead?

```
time.Time {
        wall: 13723543243807126424,
        ext: 4158612300,
        loc: *time.Location {
                name: "",
                zone: []time.zone len: 0, cap: 0, nil,
                tx: []time.zoneTrans len: 0, cap: 0, nil,
                cacheStart: 0,
                cacheEnd: 0,
                cacheZone: *time.zone nil,},}
```

How to get formatted time text in dlv debug mode?

### PR DESCRIPTION
**Pull request by aarzilli**
Fixes #999

### CODE REVIEW

File: pkg/proc/variables.go [hunk 1]
```diff
@@ -2454,3 +2458,47 @@ type constantValuesByValue []constantValue
...
+func (v *Variable) formatTime() {
+	wallv := v.fieldVariable("wall")
+	extv := v.fieldVariable("ext")
+	if wallv == nil || extv == nil || wallv.Unreadable != nil || extv.Unreadable != nil || wallv.Value == nil || extv.Value == nil {
+		return
+	}
+
+	wall, _ := constant.Uint64Val(wallv.Value)
+	ext, _ := constant.Int64Val(extv.Value)
+	_ = ext
```
Comment (polinasok) [inline pkg/proc/variables.go:2484]: Why do you need this?

File: service/api/conversions.go [hunk 2]
```diff
@@ -261,7 +261,7 @@ func VariableValueAsString(v *proc.Variable) string {
 		return convertFloatValue(v, 32)
 	case reflect.Float64:
 		return convertFloatValue(v, 64)
-	case reflect.String, reflect.Func:
+	case reflect.String, reflect.Func, reflect.Struct:
```
Comment (polinasok) [inline service/api/conversions.go:264]: All structs?

File: service/test/variables_test.go [hunk 3]
```diff
@@ -840,6 +840,9 @@ func TestEvalExpression(t *testing.T) {
 		{"ni8 << 8", false, "0", "0", "int8", nil},
 		{"ni8 >> 1", false, "-3", "-3", "int8", nil},
 		{"bytearray[0] * bytearray[0]", false, "144", "144", "uint8", nil},
+
+		// pretty printing special types
+		{"tim1", false, `time.Time(1977-05-25T18:00:00Z)…`, `time.Time(1977-05-25T18:00:00Z)…`, "time.Time", nil},
```
Comment (polinasok) [inline service/test/variables_test.go:845]: Is this always GMT+0000 timezone?
I can already imagine users wanting to see their timezone or be able to configure the timezone.

**Review by polinasok — COMMENTED** [review]
Meta question: this PR manually reconstructs things from the low-level fields. I am assuming somewhere in the standard library this is done as well when time is converted to string. That logic is probably hidden somewhere inside of unexported methods and can't be reused, but but can they at least be referenced beyond only a pointer to time.go? That could be of help in sanity checking the logic.

**Comment by aarzilli** [comment]
> Meta question: this PR manually reconstructs things from the low-level fields. I am assuming somewhere in the standard library this is done as well when time is converted to string. That logic is probably hidden somewhere inside of unexported methods and can't be reused, but but can they at least be referenced beyond only a pointer to time.go? That could be of help in sanity checking the logic.

The work here is done by copying a time.Time object read from the target into a time.Time object allocated by delve, time.go doesn't do this anywhere. The relevant comment is the one inside `type Time struct`.
We could format directly but we'd have to copy considerable more code.

**Comment by mark-pictor-csec** [comment]
@polinasok wrote
> Meta question: this PR manually reconstructs things from the low-level fields. I am assuming somewhere in the standard library this is done as well when time is converted to string. That logic is probably hidden somewhere inside of unexported methods and can't be reused, but but can they at least be referenced beyond only a pointer to time.go? That could be of help in sanity checking the logic.

Kinda tangential: I'd like to see pretty printers as starlark scripts, contributed to (and ultimately maintained in parallel with) the packages they serve. Ideally with tests to ensure they stay in sync with the go code.

This thinking insipired by `gdb` pretty-printer scripts being embedded in the distributed library, for example `libc`. I don't know where those scripts live within the libc source, but for go it'd certainly make sense to bundle them inside the package - perhaps under `time/_dlv/` or `time/_prettyprint/` or so. The underscore is just a hint for humans: `no go code under here` - though if the standard library already has another mechanism to indicate that, we should use it instead. As for actually embedding the script within binaries, that seems less necessary for go than it is for languages like C.

**Comment by aarzilli** [comment]
> I'd like to see pretty printers as starlark scripts, contributed to (and ultimately maintained in parallel with) the packages they serve

At the moment all starlark scripts are executed client side, meaning that if we did this all the pretty printing would only work on the command line client, which isn't that different from defining a `print-time` command and using it.

> I don't know where those scripts live within the libc source, but for go it'd certainly make sense to bundle them inside the package - perhaps under time/_dlv/ or time/_prettyprint/ or so

If we moved starlark evaluation to the backend we wouldn't necessarily be able to access those directories as delve's backend only has access to the executable it's debugging. Also directories that only contain non-go files are stripped from module archives, which means that you wouldn't have access to those files for any package except the packages of the main module and the standard library.

**Comment by derekparker** [comment]
I'm catching up on the discussion still a bit here however I do like the approach of using the scripting interface for the pretty printing... I know it's a lot of work and I'm not suggesting it right now but I really think it's worth a conversation and consideration to take more advantage of the scripting support.

**Comment by aarzilli** [comment]
Moving starlark evaluation to the backend is possible, there's going to be some pain points with making `dlv_command`, `read_file/write_file` and command registration work. Having starlark running in the server could also be used to create more complex breakpoint conditions.
Loading prettyprinters automatically from modules is something I'm not convinced is useful enough given the complications (also we could think about this later).
Before I port starlark to run server-side however I'd like to know three things:
1. what happens to the time.Time pretty printer? Does it become a hardcoded script?
2. are we going to add some starlark init file that gets loaded at startup by headless servers (at the moment the init file is *not* starlark and is only loaded by the frontend)
3. how does this get integrated with DAP?

**Comment by hyangah** [comment]
Let's discuss the general scripting interface in a separate issue.

As @aarzilli pointed out, the implication with DAP is unclear. Starlark scripting is too powerful (It's delve RPC-based so you can do more than just pretty printing) for DAP to embrace it without careful design IMO. It's unclear to me, who's responsible for `time.Time` pretty printing support, and how trivial it is to write in a script (unexperienced users would attempt to utilize dlv `call` command-like features which may not be desirable at this moment).

For `gdb` support, go embeds the location of a very very basic script (from GOROOT) to the `.debug_gdb_scripts` section in the binary, but I am not sure if that approach can generally work for other packages without Go tool's support.

**Comment by aarzilli** [comment]
> unexperienced users would attempt to utilize dlv call command-like features which may not be desirable at this moment

I wouldn't allow call and other commands that resume execution in this context.

**Comment by derekparker** [comment]
> Let's discuss the general scripting interface in a separate issue.

Agreed, this will be a large undertaking if we happen to go for it and would need upfront design, discussion and consideration.

**Comment by derekparker** [comment]
@aarzilli would you mind creating the issue for the scripting changes?

**Comment by aarzilli** [comment]
> There's still a few outstanding questions from @polinasok

Oops I didn't see them.

File: pkg/proc/variables.go (hunk 1 above)
Comment (aarzilli) [reply pkg/proc/variables.go]: Leftover, removed.

File: service/api/conversions.go (hunk 2 above)
Comment (aarzilli) [reply service/api/conversions.go:264]: I don't think there's a reason to be more specific.

File: service/test/variables_test.go (hunk 3 above)
Comment (aarzilli) [reply service/test/variables_test.go:845]: I'm not sure we can do any better. We could look up the time.Time.Location.name field, and try to load it with time.LoadLocation, maybe that's good enough but if the program created the timezone using time.FixedZone the name could be anything.

**Comment by aarzilli** [comment]
> @aarzilli would you mind creating the issue for the scripting changes?

Done: #2887

File: service/test/variables_test.go (hunk 3 above)
Comment (derekparker) [reply service/test/variables_test.go:845]: Possibly in the future we could let users specify this?

Comment (aarzilli) [reply service/test/variables_test.go:845]: How would they specify it?

Comment (derekparker) [reply service/test/variables_test.go:845]: I was thinking more along the lines of a custom format print command or something. Sorry, kinda thinking out loud without concrete ideas.

Comment (aarzilli) [reply service/test/variables_test.go:845]: I've changed it to load the location, if there is one and print it. If we want to go with the scripting thing then picking the location becomes part of that.

**Comment by derekparker** [comment]
Needs rebase. Also @polinasok can you do another review pass?

**Comment by aarzilli** [comment]
Rebased.

**Review by polinasok — APPROVED** [review]
Thank you for adding the location logic.

**Review by derekparker — APPROVED** [review]
LGTM

**Pull request merged**